src/Turing.jl

```diff
@@ -27,25 +28,72 @@ function setprogress!(progress::Bool)
 end
 
 # Log density function
-struct LogDensityFunction{V,M,S,C}
+struct LogDensityFunction{V,M,C}
     varinfo::V
     model::M
-    sampler::S
     context::C
 end
 
+function LogDensityFunction(
+    varinfo::DynamicPPL.AbstractVarInfo,
+    model::DynamicPPL.Model,
+    sampler::DynamicPPL.AbstractSampler,
+    context::DynamicPPL.AbstractContext
+)
+    return LogDensityFunction(varinfo, model, DynamicPPL.SamplingContext(sampler, context))
+end
+
+# Convenient for end-user.
+function LogDensityfunction(
+    model::DynamicPPL.Model,
+    varinfo::DynamicPPL.AbstractVarInfo=DynamicPPL.VarInfo(model),
+    context::DynamicPPL.AbstractContext=DynamicPPL.DefaultContext(),
+)
+    return LogDensityFunction(model, varinfo, context)
+end
```

Comment:
Should this be added at this point? IIRC it's not exported and not suggested that users use it (I think it's not even documented).

Reply:
IMO yeah; in fact, I kind of want to export it. 
Now that we're adopting the LogDensityProblems.jl interface more aggressively, I also find myself using it _very_ often. I also very much like it because most users will find the "full" description necessary to run the model, i.e. `model` + `varinfo` + `context` too "scary". If we can say "if you want to play around with your model, you can convert it to a simple thing which is compatible with LogDensityProblems", I think that's quite nice.

Reply:
I like it as well and I think we want to expose it somehow. My main questions are just whether the right time to do is during the refactoring and whether we want to expose it via a different API. For instance, we could also use something like
```julia
logjoint(model, ::Type{T}=VarInfo) where {T<:AbstractVarInfo} = LogDensityFunction(model, T(model), DefaultContext())
logprior(model, ::Type{T}=VarInfo) where {T<:AbstractVarInfo} = LogDensityFunction(model, T(model), PriorContext())
loglikelihood(model, ::Type{T}=VarInfo) where {T<:AbstractVarInfo} = LogDensityFunction(model, T(model), LikelihoodContext())
```
and keep `LogDensityFunction` an internal detail. Maybe that would be even easier for users.

Reply:
Though I do like this interface, I feel like we still need to inform the end-user about it and might as well export it given that they'll be seeing it "a lot"?

Reply:
I'm not sure if they have to know about `LogDensityFunctions` at all if `logjoint` etc would be the API. The only relevant point for thr uaers would be (and should be documented) that these functions return something that follows the LogDensityProblems interface. Fields, names etc. should not matter for users.

Reply:
A bit similar to how users don't know (and don't have to know) what `truncated` returns. Similarly, `logjoint` would make it easier to use the same API for other models for which we might need something different than `LogDensityFunctions`.

Reply:
I have seen some cases where the user manually constructs these `LogDensityFunctions`.

 It is an intuitive interface for working with models, compared to treating them as a somewhat black-box object where users can only interact with a set of APIs. It allows users to explore new sampling algorithms, for instance, or plug a Turing model inside a large manually coded models.

Reply:
I think @devmotion agrees with this, but he's just saying that generally the user shouldn't need to use the constructor itself, but instead use the `logjoint`, etc. methods for constructing it. 

I think I'll agree with not exporting it for now, but instead define these "nicer" constructors just because exporting a new thing is non-breaking but removing an export _is_ breaking, hence we can always just export it later if we feel like it's useful.

Reply:
I don't question its usefulness (actually I think it's very useful and not surprised that it's used even though it's not documented) but solely the two points mentioned: whether we should add to the API in a PR that otherwise only performs internal refactoring and whether a more generic API without exposing `LogDensityFunctions` directly would be better.

Reply:
Btw, is there a reason why we prefer something like:
```julia
logjoint(model::Model, ::Type{T} = VarInfo) = LogDensityFunction(VarInfo(model), model, ...)
```
vs.
```julia
logjoint(model::Model, varinfo::AbstractVarInfo = VarInfo(model)) = LogDensityFunction(varinfo, model, ...)
```
/

Reply:
This constructor is actually quite a bit annoying as it doesn't let you re-use an existing `AbstractVarInfo` :confused:

Reply:
It was just the first thing I thought of, maybe there are better approaches. But maybe that's also an indication that a separate PR and discussion about the API could be uaeful.

Reply:
Could always add `make_logjoint`, etc.?

Reply:
I agree with a separate PR, but worried it's going to delay exposing the LogDensityProblems.jl-functionality to the users. But I'll make separate PR for now, and if it takes too long we can just export `LogDensityFunction` :shrug:

Reply:
Or something like `logjointof` etc., similar to `(log)densityof`.

On a second thought though, I think generally (and particularly when dealing with other model type) it would be nice if there would be some curried form of `logjoint` etc. that gives you a function that you can evaluate for different variables and that would be consistent with the exiating `logjoint` etc. And generally there one would only want to provide the variables when evaluating the curried form. Maybe one could add a keyword argument for the LogDensityFunction case (similar to e.g. how one can adjust behaviour of AD backends in LogDensityProblemsAD by providing an example of the values that log density function will be evaluated with). But to summarize, to me the desire to provide `VarInfo(model)` when constructing the LogDensityFunction seems very specific to this case (and maybe not something many users would want to do?) but not something you would want to do for every kind of models, and hence maybe it should not be first-class part of a generic API but a bit more specific to this case such as maybe a keyword argument.